### ISSUE TITLE
It is not possible to detect if nested contexts are transactional.

### ISSUE THREAD

**Issue by jongillham** [qedus/nds#37]
Take the following situation where I create a custom context:

```go
type myContext struct {
    appengine.Context
} 
```

If I do this:
```go
nds.RunInTransaction(c, func(tc appengine.Context) error {
    mc := &myContext{tc}

    // The following calls to the NDS API will not be executed
    // the transactional code path.
    nds.Get(mc, key, val)
    nds.Put(mc, key, val)
    nds.Delete(mc, key)
}, nil) 
```

There is no way of the `nds.Get/Put/Delete` functions knowing that the `mc` context is a transactional context in the current API. Note that if the context is not wrapped then `nds.Get/Put/Delete` works as expected.

The only nice way I can see around this subtle issue is to *change* the NDS API as shown in this context branch: https://github.com/qedus/nds/tree/context

I wrap `appengine.Context`s a lot to add extra functionality that must be bound to a single context/request. The changes I propose to the API will allow me to continue doing this without fear that I am not in a transaction when I think I am.

There is some more discussion here: https://groups.google.com/forum/#!topic/golang-nuts/xrxjgCGzUfs

**Comment by derekperkins** [comment]
This would only be necessary when you want to support transactions, correct? If that's the case, I can definitely get behind this.

**Comment by jongillham** [comment]
@derekperkins the issue only has the potential to arise when you are in a transaction. However the solution I propose (https://github.com/qedus/nds/tree/context) would affect the signatures of the whole API. Instead of doing `nds.Get(c, key, value)` you would need to do `c.Get(key, value)`. I am guessing you would not be happy with this?

There might be a happy middle ground though where only code that runs in a transaction must change...

**Comment by derekperkins** [comment]
Everything Google is doing is migrating towards using https://godoc.org/golang.org/x/net/context, including the appengine package. My vote would be for you to not introduce your own context, but instead use this new one and pass it through as the first parameter of the function. That's at least the most future proof way that I can think of.

I don't have a great solution for the existing appengine package though, since it doesn't have a `Value` function you can hook into.

**Comment by jongillham** [comment]
That is an excellent suggestion. I am going to refactor the context branch with your suggestions.

**Comment by jongillham** [comment]
Fixed in commit 2d34590.
